File: NetworkSensorAppJava/app/src/main/java/com/networksensor/app/model/SensorData.java
package com.networksensor.app.model;

import android.os.Parcel;
import android.os.Parcelable;
import java.util.Arrays;

public class SensorData implements Parcelable {
    private SensorType sensorType;
    private float[] values;
    private int accuracy;
    private long timestamp;

    public SensorData() {
        this.timestamp = System.currentTimeMillis();
    }

    public SensorData(SensorType sensorType, float[] values, int accuracy) {
        this.sensorType = sensorType;
        this.values = values != null ? values.clone() : new float[0];
        this.accuracy = accuracy;
        this.timestamp = System.currentTimeMillis();
    }

    protected SensorData(Parcel in) {
        sensorType = SensorType.valueOf(in.readString());
        values = in.createFloatArray();
        accuracy = in.readInt();
        timestamp = in.readLong();
    }

    public static final Creator<SensorData> CREATOR = new Creator<SensorData>() {
        @Override
        public SensorData createFromParcel(Parcel in) {
            return new SensorData(in);
        }

        @Override
        public SensorData[] newArray(int size) {
            return new SensorData[size];
        }
    };

    // Getters and Setters
    public SensorType getSensorType() {
        return sensorType;
    }

    public void setSensorType(SensorType sensorType) {
        this.sensorType = sensorType;
    }

    public float[] getValues() {
        return values != null ? values.clone() : new float[0];
    }

    public void setValues(float[] values) {
        this.values = values != null ? values.clone() : new float[0];
    }

    public int getAccuracy() {
        return accuracy;
    }

    public void setAccuracy(int accuracy) {
        this.accuracy = accuracy;
    }

    public long getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(long timestamp) {
        this.timestamp = timestamp;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(sensorType.name());
        dest.writeFloatArray(values);
        dest.writeInt(accuracy);
        dest.writeLong(timestamp);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        SensorData that = (SensorData) o;
        return accuracy == that.accuracy &&
                timestamp == that.timestamp &&
                sensorType == that.sensorType &&
                Arrays.equals(values, that.values);
    }

    @Override
    public int hashCode() {
        int result = sensorType != null ? sensorType.hashCode() : 0;
        result = 31 * result + Arrays.hashCode(values);
        result = 31 * result + accuracy;
        result = 31 * result + (int) (timestamp ^ (timestamp >>> 32));
        return result;
    }

    public enum SensorType {
        ACCELEROMETER,
        GYROSCOPE,
        MAGNETOMETER,
        GRAVITY,
        LINEAR_ACCELERATION,
        ROTATION_VECTOR,
        ORIENTATION,
        PRESSURE,
        TEMPERATURE,
        HUMIDITY,
        LIGHT,
        PROXIMITY,
        STEP_COUNTER,
        HEART_RATE
    }
}

// Location Data class
class LocationData implements Parcelable {
    private double latitude;
    private double longitude;
    private double altitude;
    private float accuracy;
    private float speed;
    private float bearing;
    private String provider;
    private long timestamp;

    public LocationData() {
        this.timestamp = System.currentTimeMillis();
    }

    public LocationData(double latitude, double longitude, double altitude, float accuracy,
                       float speed, float bearing, String provider) {
        this.latitude = latitude;
        this.longitude = longitude;
        this.altitude = altitude;
        this.accuracy = accuracy;
        this.speed = speed;
        this.bearing = bearing;
        this.provider = provider;
        this.timestamp = System.currentTimeMillis();
    }

    protected LocationData(Parcel in) {
        latitude = in.readDouble();
        longitude = in.readDouble();
        altitude = in.readDouble();
        accuracy = in.readFloat();
        speed = in.readFloat();
        bearing = in.readFloat();
        provider = in.readString();
        timestamp = in.readLong();
    }

    public static final Creator<LocationData> CREATOR = new Creator<LocationData>() {
        @Override
        public LocationData createFromParcel(Parcel in) {
            return new LocationData(in);
        }

        @Override
        public LocationData[] newArray(int size) {
            return new LocationData[size];
        }
    };

    // Getters and Setters
    public double getLatitude() { return latitude; }
    public void setLatitude(double latitude) { this.latitude = latitude; }
    
    public double getLongitude() { return longitude; }
    public void setLongitude(double longitude) { this.longitude = longitude; }
    
    public double getAltitude() { return altitude; }
    public void setAltitude(double altitude) { this.altitude = altitude; }
    
    public float getAccuracy() { return accuracy; }
    public void setAccuracy(float accuracy) { this.accuracy = accuracy; }
    
    public float getSpeed() { return speed; }
    public void setSpeed(float speed) { this.speed = speed; }
    
    public float getBearing() { return bearing; }
    public void setBearing(float bearing) { this.bearing = bearing; }
    
    public String getProvider() { return provider; }
    public void setProvider(String provider) { this.provider = provider; }
    
    public long getTimestamp() { return timestamp; }
    public void setTimestamp(long timestamp) { this.timestamp = timestamp; }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeDouble(latitude);
        dest.writeDouble(longitude);
        dest.writeDouble(altitude);
        dest.writeFloat(accuracy);
        dest.writeFloat(speed);
        dest.writeFloat(bearing);
        dest.writeString(provider);
        dest.writeLong(timestamp);
    }
}

File: NetworkSensorApp/app/src/main/java/com/networksensor/app/data/model/SensorData.kt
package com.networksensor.app.data.model

import android.os.Parcelable
import kotlinx.parcelize.Parcelize

@Parcelize
data class SensorReading(
    val sensorType: SensorType,
    val values: FloatArray,
    val accuracy: Int,
    val timestamp: Long = System.currentTimeMillis()
) : Parcelable {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as SensorReading

        if (sensorType != other.sensorType) return false
        if (!values.contentEquals(other.values)) return false
        if (accuracy != other.accuracy) return false
        if (timestamp != other.timestamp) return false

        return true
    }

    override fun hashCode(): Int {
        var result = sensorType.hashCode()
        result = 31 * result + values.contentHashCode()
        result = 31 * result + accuracy
        result = 31 * result + timestamp.hashCode()
        return result
    }
}

enum class SensorType {
    ACCELEROMETER,
    GYROSCOPE,
    MAGNETOMETER,
    GRAVITY,
    LINEAR_ACCELERATION,
    ROTATION_VECTOR,
    ORIENTATION,
    PRESSURE,
    TEMPERATURE,
    HUMIDITY,
    LIGHT,
    PROXIMITY,
    STEP_COUNTER,
    HEART_RATE
}

@Parcelize
data class LocationData(
    val latitude: Double,
    val longitude: Double,
    val altitude: Double,
    val accuracy: Float,
    val speed: Float,
    val bearing: Float,
    val provider: String,
    val timestamp: Long = System.currentTimeMillis()
) : Parcelable

@Parcelize
data class DeviceMotion(
    val accelerometer: FloatArray,
    val gyroscope: FloatArray,
    val magnetometer: FloatArray,
    val gravity: FloatArray,
    val linearAcceleration: FloatArray,
    val rotationVector: FloatArray,
    val timestamp: Long = System.currentTimeMillis()
) : Parcelable {
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as DeviceMotion

        if (!accelerometer.contentEquals(other.accelerometer)) return false
        if (!gyroscope.contentEquals(other.gyroscope)) return false
        if (!magnetometer.contentEquals(other.magnetometer)) return false
        if (!gravity.contentEquals(other.gravity)) return false
        if (!linearAcceleration.contentEquals(other.linearAcceleration)) return false
        if (!rotationVector.contentEquals(other.rotationVector)) return false
        if (timestamp != other.timestamp) return false

        return true
    }

    override fun hashCode(): Int {
        var result = accelerometer.contentHashCode()
        result = 31 * result + gyroscope.contentHashCode()
        result = 31 * result + magnetometer.contentHashCode()
        result = 31 * result + gravity.contentHashCode()
        result = 31 * result + linearAcceleration.contentHashCode()
        result = 31 * result + rotationVector.contentHashCode()
        result = 31 * result + timestamp.hashCode()
        return result
    }
}

@Parcelize
data class EnvironmentalData(
    val temperature: Float?,
    val humidity: Float?,
    val pressure: Float?,
    val lightLevel: Float?,
    val proximity: Float?,
    val timestamp: Long = System.currentTimeMillis()
) : Parcelable